flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetUnion.scala

```diff
@@ -69,16 +73,23 @@ class DataSetUnion(
       rows + metadata.getRowCount(child)
     }
 
-    planner.getCostFactory.makeCost(rowCnt, 0, 0)
+    planner.getCostFactory.makeCost(
+      rowCnt,
+      if (all) 0 else rowCnt,
+      if (all) 0 else rowCnt)
   }
 
   override def translateToPlan(
       tableEnv: BatchTableEnvironment,
       expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
 
-    val leftDataSet = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
-    val rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
-    leftDataSet.union(rightDataSet).asInstanceOf[DataSet[Any]]
+    val leftDataSet = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
+    val rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
+    if (all) {
+      leftDataSet.union(rightDataSet).asInstanceOf[DataSet[Any]]
+    } else {
+      leftDataSet.union(rightDataSet).distinct().asInstanceOf[DataSet[Any]]
```

Comment:
Oh, yes. Completely forgot about that rule... 😊 
So, we already supported the non-all union for SQL. Only the Table API was missing the `union()` method.
I think there are two ways to continue: 
- remove the `UnionToDistinctRule` from `FlinkRuleSets`
- revert the changes on `DataSetUnion` (except of pushing down the `expectedType`) and `DataSetUnionRule`.

I am fine either ways.